snapshot-service/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+mod handlers;
+mod routes;
+
+use jormungandr_lib::{crypto::account::Identifier, interfaces::Value};
+pub use routes::filter;
+use sled::Transactional;
+use std::cmp::max;
+use voting_hir::VoterHIR;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    DbError(#[from] sled::Error),
+
+    #[error(transparent)]
+    DbTxError(#[from] sled::transaction::TransactionError),
+
+    #[error("internal error")]
+    InternalError,
+}
+
+pub type Tag = String;
+type Group = String;
+
+#[derive(Clone)]
+pub struct SharedContext {
+    db: sled::Db,
+}
+
+impl SharedContext {
+    #[tracing::instrument(skip(self))]
+    pub fn get_voting_power(
+        &self,
+        tag: &str,
+        id: &Identifier,
+    ) -> Result<Option<Vec<(Group, Value)>>, Error> {
+        let entries = self.db.open_tree("entries")?;
+        let tags = self.db.open_tree("tags")?;
+
+        let tag = if let Some(tag) = tags.get(tag)? {
+            tag
+        } else {
+            return Ok(None);
+        };
+
+        let key = {
+            let mut key = [0u8; 32 + 4];
+
+            let (tag_part, id_part) = key.split_at_mut(4);
+            tag_part.copy_from_slice(&*tag);
+
+            id_part.copy_from_slice(id.as_ref().as_ref());
+
+            key
+        };
+
+        let key_len = key.len();
+
+        let mut result: Vec<(Group, Value)> = vec![];
+
+        for entries in entries.range(key.clone()..) {
+            let (k, v) = entries?;
+
+            if k[0..key_len] > key[..] {
+                break;
+            }
+
+            let group =
+                String::from_utf8(k[key_len..].to_vec()).map_err(|_| Error::InternalError)?;
+
+            let voting_power = Value::from(u64::from_be_bytes(v.as_ref().try_into().unwrap()));
+
+            result.push((group, voting_power));
+        }
+
+        Ok(Some(result))
+    }
+
+    pub fn get_tags(&self) -> Result<Vec<Tag>, Error> {
+        let tags = self.db.open_tree("tags")?;
+
+        let mut result = vec![];
+        for entries in tags.iter() {
+            let (tag, _) = entries?;
+            result.push(String::from_utf8(tag.to_vec()).map_err(|_| Error::InternalError)?);
+        }
+
+        Ok(result)
+    }
+}
+
+// do NOT implement/derive Clone for this. The implementation of update relies on &mut self and the
+// split in a reader type and a writer type is to enforce a single writer.
+pub struct UpdateHandle {
+    _db: sled::Db,
+    tags: sled::Tree,
+    entries: sled::Tree,
+    next_tag_id: u32,
+}
+
+impl UpdateHandle {
+    pub fn new(db: sled::Db) -> Result<Self, Error> {
+        let tags = db.open_tree("tags")?;
+        let entries = db.open_tree("entries")?;
+
+        let mut next_tag_id = None;
+
+        for kv in tags.iter() {
+            let (_, v) = kv?;
+
+            let id = u32::from_be_bytes(v.as_ref().try_into().unwrap());
+
+            next_tag_id = next_tag_id.map(|c| max(c, id)).or_else(|| Some(id));
```

Comment:
this could be rewritten as something like `tags.iter()[..].map(|(k, v)| u32::from_be_bytes(v.as_ref().try_into().unwrap())).max().unwrap_or(0)`